### PR TITLE
Added transitions for Presence Event Engine.

### PR DESCRIPTION
**Pull request by marcin-cebo**

### CODE REVIEW

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/SubscribeSubscribeEventEngineTest.kt [hunk 1]
```diff
@@ -2,7 +2,7 @@ package com.pubnub.api.subscribe.eventengine
 
 import org.junit.jupiter.api.Test
 
-class SubscribeEventEngineTest {
+class SubscribeSubscribeEventEngineTest {
```
Comment (kleewho) [inline src/test/kotlin/com/pubnub/api/subscribe/eventengine/SubscribeSubscribeEventEngineTest.kt:5]: The replace went too far

**Review by kleewho — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/pubnub/api/presence/eventengine/effect/PresenceEffectInvocation.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.pubnub.api.presence.eventengine.effect
+
+import com.pubnub.api.eventengine.Cancel
+import com.pubnub.api.eventengine.EffectInvocation
+import com.pubnub.api.eventengine.EffectInvocationType
+import com.pubnub.api.eventengine.Managed
+
+sealed class PresenceEffectInvocation(override val type: EffectInvocationType) : EffectInvocation {
+    override val id: String = "any" // todo does value of this matter?
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/effect/PresenceEffectInvocation.kt:9]: Ok. I will make changes here and in SubscribeEffectInvocation class.

File: src/main/kotlin/com/pubnub/api/presence/eventengine/state/PresenceState.kt [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        override fun transition(event: PresenceEvent): Pair<PresenceState, List<PresenceEffectInvocation>> {
+            return when (event) {
+                is PresenceEvent.LeftAll -> {
+                    transitionTo(HearbeatInactive, PresenceEffectInvocation.Leave(channels, channelGroups))
+                }
+                is PresenceEvent.Disconnect -> {
+                    transitionTo(HeartbeatStopped(channels, channelGroups), PresenceEffectInvocation.Leave(channels, channelGroups))
+                }
+                is PresenceEvent.Joined -> {
+                    transitionTo(Heartbeating(event.channels, event.channelGroups))
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/state/PresenceState.kt:39]: I am wondering why there is discrepancy between how Subscribe EE handles this. I will modify code based on info from you hopefully find the answer.

File: src/main/kotlin/com/pubnub/api/presence/eventengine/state/PresenceState.kt [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                    transitionTo(HearbeatInactive)
+                }
+                is PresenceEvent.Joined -> {
+                    transitionTo(HeartbeatStopped(event.channels, event.channelGroups))
+                }
+                is PresenceEvent.Left -> {
+                    transitionTo(HeartbeatStopped(event.channels, event.channelGroups))
+                }
+                is PresenceEvent.StateSet -> {
+                    transitionTo(Heartbeating(event.channels, event.channelGroups))
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/state/PresenceState.kt:118]: Yes. Good catch.

File: src/test/kotlin/com/pubnub/api/subscribe/eventengine/SubscribeSubscribeEventEngineTest.kt (hunk 1 above)
Comment (marcin-cebo) [reply src/test/kotlin/com/pubnub/api/subscribe/eventengine/SubscribeSubscribeEventEngineTest.kt:5]: Yes. Good catch.

File: src/main/kotlin/com/pubnub/api/presence/eventengine/transition/Transition.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.pubnub.api.presence.eventengine.transition
+
+import com.pubnub.api.presence.eventengine.effect.PresenceEffectInvocation
+import com.pubnub.api.presence.eventengine.event.PresenceEvent
+import com.pubnub.api.presence.eventengine.state.PresenceState
+
+internal fun transition(state: PresenceState, presenceEvent: PresenceEvent): Pair<PresenceState, List<PresenceEffectInvocation>> = state.transition(presenceEvent)
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/transition/Transition.kt:7]: Ok. For now I will leave it as is. Initially I wanted to do so but I had hard time to do this.

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/transition/Transition.kt:7]: This I think should do the trick 


```suggestion
internal fun <EV : Event, EF : EffectInvocation, S : State<EF, EV, S>>  transition(state: S, event: EV): Pair<S, List<EF>> = state.transition(event)
```

[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/presence/eventengine/transition/Transition.kt]: Ok. I will try that.

**Review by kleewho — APPROVED** [review]

**Pull request merged**